### PR TITLE
[doc]Spark writer

### PR DESCRIPTION
**Pull request by Amber1990Zhang**
Code pr #1210.

### CODE REVIEW

File: docs/manual-CN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+
+```bash
+git clone xxxxx
+mvn compile package
+```
+
+### 2.2 oss 下载
+
+```bash
+wget xxx
```
Comment (darionyaphet) [inline docs/manual-CN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: `xxx` ?

**Comment by Amber1990Zhang** [comment]
> what is `xxx` ? we should specified the path

Thanks for your review. So what is the exact path? Will you show me?

File: docs/manual-CN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md [hunk 2]
```diff
@@ -0,0 +1,345 @@
+# Spark Writer
+
+## 1. 概述
+
+Spark Writer 是 Nebula Graph 基于 Spark 的分布式数据导入工具，能够将多种数据源中的数据转化为图的点和边并批量导入到图数据库中。目前支持的数据源有：
+
+* HDFS 文件，支持 Parquet、JSON、ORC 和 CSV 格式的文件
```
Comment (wilsonyou) [inline docs/manual-CN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: HDFS 文件，Parquet、JSON、ORC 和 CSV 格式的文件

Comment (darionyaphet) [reply docs/manual-CN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: HDFS 其实是系统 Parquet、JSON、ORC 和 CSV是文件格式 「感觉说不清楚」

File: docs/manual-CN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+
+```text
+{"source":100,"target":101,"likeness":95,"ranking":2}
+{"source":101,"target":100,"likeness":95,"ranking":1}
+{"source":101,"target":102,"likeness":90,"ranking":3}
+```
+
+#### 3.2.3 Geo 数据支持
+
+Spark Writer 支持 Geo 数据导入，Geo 数据用 latitude 与 longitude 字段描述经纬度，数据类型为 double。
```
Comment (wilsonyou) [inline docs/manual-CN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: **latitude**

File: docs/manual-CN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+      retry: 3
+    }
+  }
+
+  # 标签处理
+  tags: {
+
+    # 从 HDFS 文件加载数据， 此处数据类型为 Parquet
+    # tag 名称为 tag name 0
+    #  HDFS Parquet 文件的中的 field 0，field 1，field 2 将写入 tag name 0
```
Comment (wilsonyou) [inline docs/manual-CN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: field 0、

File: docs/manual-CN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+      }
+      vertex: vertex id field
+    }
+  }
+
+  # 边处理
+  edges: {
+    # 从 HDFS 加载数据，数据类型为 JSON
+    # 边名称为 edge name 0
+    # HDFS JSON 文件中的 field 0，field 1，field 2 将被写入 edge name 0
```
Comment (wilsonyou) [inline docs/manual-CN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: ditto

File: docs/manual-CN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+| nebula.connection.timeout | 3000 | 否 | Thrift 连接超时时间 |
+| nebula.connection.retry | 3 | 否 | Thrift 连接重试次数 |
+| nebula.execution.retry | 3 | 否 | nGQL 语句执行重试次数 |
+
+#### 3.3.3 tags 和 edges 映射信息
+
+tag 和 edge 映射的选项比较类似。下面先介绍相同的选项，再分别介绍 `tag 映射`和 `edge 映射`的特有选项。
+
+* **相同的选项**
+  * `type` 指定上文中提到的数据类型，目前支持 “Parquet”, "JSON", "ORC" 和 “CSV” ，大小写不敏感，必填
```
Comment (wilsonyou) [inline docs/manual-CN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: 、、、

File: docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+
+### 2.1 From Source Code
+
+```bash
+git clone https://github.com/vesoft-inc/nebula.git
+cd nebula/src/tools/spark-sstfile-generator
+mvn compile package
+```
+
+### 2.2 oss Download
```
Comment (wilsonyou) [inline docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: Download OSS

File: docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+
+* Edge with rank, take edge _**follow**_ as example
+
+```text
+{"source":100,"target":101,"likeness":95,"ranking":2}
+{"source":101,"target":100,"likeness":95,"ranking":1}
+{"source":101,"target":102,"likeness":90,"ranking":3}
+```
+
+#### 3.2.3 Supporting for Geo Data
```
Comment (wilsonyou) [inline docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: Support for

File: docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+
+```text
+age,id,name
+42,100,Tim Duncan
+36,101,Tony Parker
+```
+
+##### Database
+
+Spark Writer supports database as the data source, only HIVE is available for now.
```
Comment (wilsonyou) [inline docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: Spark Writer supports database as the data source, and only HIVE is available now.

File: docs/manual-CN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md (hunk 2 above)
Comment (Amber1990Zhang) [reply docs/manual-CN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: HDFS 文件，包括 Parquet、JSON、ORC 和 CSV 格式的文件

File: docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+```text
+col_name             data_type           comment
+id                   int
+name                 string
+age                  int
+```
+
+### 3.3 Write Configuration Files
+
+The configuration files consists of Spark related information, Nebula related information, and tags mapping and edges mapping blocks. Spark information is configured with the associated parameters running Spark. Nebula information is configured with information such as user name and password to connect Nebula. Tags mapping and edges mapping correspond to the input source mapping of multiple tag/edges respectively, describing the basic information like each tag/edge's data source. It's possible that different tag/edge come from different daa sources.
```
Comment (wilsonyou) [inline docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: consist of

Comment (wilsonyou) [inline docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: daa should be data

File: docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md [hunk 11]
```diff
@@ -0,0 +1,344 @@
...
+  * `ranking` is an optional field that specifies a column as the edge ranking column when the inserted edge has a ranking value
+
+#### 3.3.4 Data Source Mapping
+
+* **HDFS Parquet Files**
+  * `type` specifies the input source type. When it is parquet, it's a case insensitive required field
+  * `path` specifies the HDFS file directory, it's a required field that must be the absolute directory
+* **HDFS JSON Files**
+  * `type` specifies the type of the input source. When it is JSON, it's a case insensitive required field
+  * `path` specifies the HDFS file directory, it's a required field that must be absolute directory
```
Comment (wilsonyou) [inline docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: It is a required field

File: docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+ --master ${MASTER-URL} \
+ ${SPARK_WRITER_JAR_PACKAGE} -c conf/test.conf -h -d
+```
+
+Parameter descriptions:
+
+| Abbreviation | Required | Default | Description | Example |
+| --- | --- | --- | --- | --- |
+| --class | yes | / | Specify the program's main class |  |
+| --master | yes | / | Specify spark cluster master url, refer to [master urls](https://spark.apache.org/docs/latest/submitting-applications.html#master-urls) for detail | e.g. `spark://23.195.26.187:7077` |
```
Comment (wilsonyou) [inline docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: . Refer to

File: docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md [hunk 13]
```diff
@@ -0,0 +1,344 @@
...
+| -c / --config  | yes | / | The configuration file path in the context |  |
+| -h / --hive | no | false | Used to specify whether to support Hive |  |
+| -d / --directly | no | false | true for console insertion; <br>false for sst import (TODO) |  |
```
Comment (wilsonyou) [inline docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: True

File: docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md (hunk 12 above)
Comment (wilsonyou) [inline docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: Refer to

File: docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md [hunk 14]
```diff
@@ -0,0 +1,344 @@
...
+* **HIVE CSV Files**
+  * `type` specifies the input source type. When it is CSV, it's a case insensitive required field
+  * `path` specifies the HDFS file directory, it's a required field that must be the absolute directory
+* **HIVE**
+  * `type` specifies the input source type. When it is HIVE, it's a case insensitive required field
+  * `exec` is a required field that specifies the HIVE executed query
+
+### 3.4 Import Data
+
+Input data use the following command:
```
Comment (wilsonyou) [inline docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: The input data uses the ...

File: docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md [hunk 15]
```diff
@@ -0,0 +1,344 @@
...
+
+* **HDFS Parquet Files**
+  * `type` specifies the input source type. When it is parquet, it's a case insensitive required field
+  * `path` specifies the HDFS file directory, it's a required field that must be the absolute directory
+* **HDFS JSON Files**
+  * `type` specifies the type of the input source. When it is JSON, it's a case insensitive required field
+  * `path` specifies the HDFS file directory, it's a required field that must be absolute directory
+* **HIVE ORC Files**
+  * `type` specifies the input source type. When it is ORC, it's a case insensitive required field
+  * `path` specifies the HDFS file directory, it's a required field that must be absolute directory
```
Comment (wilsonyou) [inline docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: must be the absolute...

File: docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md [hunk 16]
```diff
@@ -0,0 +1,344 @@
...
+| nebula.execution.retry | 3 | no | nGQL execution retry times |
+
+#### 3.3.3 Mapping of Tags and Edges
+
+The options for tag and edge mapping are very similar. The following describes the same options first, and then introduces the unique options of `tag mapping` and `edge mapping`.
+
+* **Same Options**
+  * `type` is a case insensitive required field that specifies data type in the context, currently supports Parquet, JSON, ORC and CSV
+  * `path` is applied to HDFS data source and specifies the absolute path of HDFS file or directory, it's a required field when the type is HDFS
+  * `exec` is applied to Hive data source, it's a required filed when the query type is HIVE
```
Comment (wilsonyou) [inline docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: It is a required

File: docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md [hunk 17]
```diff
@@ -0,0 +1,344 @@
...
+| nebula.connection.timeout | 3000 | no | Thrift timeout |
+| nebula.connection.retry | 3 | no | Thrift retry times |
+| nebula.execution.retry | 3 | no | nGQL execution retry times |
+
+#### 3.3.3 Mapping of Tags and Edges
+
+The options for tag and edge mapping are very similar. The following describes the same options first, and then introduces the unique options of `tag mapping` and `edge mapping`.
+
+* **Same Options**
+  * `type` is a case insensitive required field that specifies data type in the context, currently supports Parquet, JSON, ORC and CSV
```
Comment (wilsonyou) [inline docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: currently support for

File: docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md [hunk 18]
```diff
@@ -0,0 +1,344 @@
...
+
+#### 3.3.1 Spark Properties
+
+The following table gives some example properties, all of which can be found in [Spark Available Properties](http://spark.apache.org/docs/latest/configuration.html#available-properties).
+
+| Field | Default | Required | Description |
+| --- | --- | --- | --- |
+| spark.app.name | Spark Writer | No | The name of your application |
+| spark.driver.cores | 1 | No | Number of cores to use for the driver process, only in cluster mode. |
+| spark.driver.maxResultSize | 1G | No | Limit of total size of serialized results of all partitions for each Spark action (e.g. collect) in bytes. Should be at least 1M, or 0 for unlimited. |
```
Comment (wilsonyou) [inline docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: Should be ... It must be at least

File: docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md (hunk 14 above)
Comment (Amber1990Zhang) [reply docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: I'd rather an imperative form here. Changed to 
Input data with the following command:

File: docs/manual-CN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md [hunk 19]
```diff
@@ -0,0 +1,345 @@
...
+
+### 3.1 构图
+
+构图请参考[快速试用](../../../../../1.overview/2.quick-start/1.get-started.md)中的示例构图。
+
+### 3.2 数据示例
+
+#### 3.2.1 点
+
+顶点数据文件由一行一行的数据组成，一般一行表示一个点，其中一列为点的 ID，此列的名称将在后文的映射文件中指定。其他列为点属性。下面以 JSON 格式为例进行说明。
```
Comment (jievince) [inline docs/manual-CN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: > 其中一列
感觉指代不明确，说成每行的第一个键值对，是不是会好一点？

File: docs/manual-CN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md [hunk 20]
```diff
@@ -0,0 +1,345 @@
...
+| --class | yes |  | 指定程序主类 |  |
+| --master | yes |  | 指定spark cluster master url，请参见 [master-urls](https://spark.apache.org/docs/latest/submitting-applications.html#master-urls) | e.g. `spark://23.195.26.187:7077` |
+| -c / --config  | yes |  | 上文所编写的配置文件路径 |  |
+| -h / --hive | no | false | 用于指定是否支持 Hive |  |
+| -d / --directly | no | false | true 为客户端方式插入；<br>false 为 sst 方式导入 (TODO) |  |
+| -D / --dry | no | false | 检查配置文件是否正确 |  |
+
+## 4. 性能测试结果
+
+三个节点，写 1 亿条数据，每条数据三个字段，每个批 64 条记录，用时四分钟左右。
```
Comment (whitewum) [inline docs/manual-CN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: 三台物理机 (56 核，250G 内存，万兆网，SSD），
写 1 亿条数据（每条数据三个字段，每个batch 64 条记录），
用时 4 分钟（40万条/秒）

File: docs/manual-CN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md [hunk 21]
```diff
@@ -2,47 +2,51 @@
...
-Spark Writer 是 Nebula Graph 基于 Spark 的分布式数据导入工具，能够将多种数据源中的数据转化为图的点和边并批量导入到图数据库中。目前支持的数据源有：
+Spark Writer 是 Nebula Graph 基于 Spark 的分布式数据导入工具，能够将多种数据仓库中的数据转化为图的点和边，并批量导入到图数据库中。目前支持的数据仓库有：
 
 * HDFS，包括 Parquet、JSON、ORC 和 CSV 格式的文件
 * HIVE
 
-Spark Writer 支持同时导入多个 tag、edge，支持不同 tag/edge 配置不同的数据源。
+Spark Writer 支持并发导入多个 tag、edge，支持不同 tag/edge 配置不同的数据仓库。
 
 ## 2. 获取 Spark Writer
```
Comment (wilsonyou) [inline docs/manual-CN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: 您可以通过以下两种方式获取Spark Writer:
1. 
2.

File: docs/manual-CN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md [hunk 22]
```diff
@@ -2,47 +2,51 @@
...
 
-### 2.1 编译源码
+### 2.1a 编译源码
```
Comment (wilsonyou) [inline docs/manual-CN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: 2.1 is fine. I think a is not needed

File: docs/manual-CN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md [hunk 23]
```diff
@@ -2,47 +2,51 @@
...
 ```bash
 git clone https://github.com/vesoft-inc/nebula.git
 cd nebula/src/tools/spark-sstfile-generator
 mvn compile package
 ```
 
-### 2.2 oss 下载
+或者直接下载
+
+### 2.1b 从云存储 OSS 下载
```
Comment (wilsonyou) [inline docs/manual-CN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: Just 2.1. b is not needed

File: docs/manual-CN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md [hunk 24]
```diff
@@ -2,47 +2,51 @@
...
 
 ## 3. 使用流程
 
 基本流程分为以下几步：
 
-* 在 Nebula 中创建图模型，构图
-* 编写数据文件
-* 编写输入源映射文件
-* 导入数据
+1. 在 Nebula 中创建图模型，构图
```
Comment (wilsonyou) [inline docs/manual-CN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: Should we use **Nebula Graph**

File: docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md [hunk 25]
```diff
@@ -0,0 +1,349 @@
...
+| --class | yes | / | Specify the program's main class |  |
+| --master | yes | / | Specify spark cluster master url. Refer to [master urls](https://spark.apache.org/docs/latest/submitting-applications.html#master-urls) for detail | e.g. `spark://23.195.26.187:7077` |
+| -c / --config  | yes | / | The configuration file path in the context |  |
+| -h / --hive | no | false | Used to specify whether to support Hive |  |
+| -d / --directly | no | false | True for console insertion; <br>false for sst import (TODO) |  |
+| -D / --dry | no | false | Check if the configuration file is correct |  |
+
+## Performance
+
+It takes about four minutes (i.e. 400 thousand rows per second) to input 100 million rows (each row contains three fields, each batch contains 64 rows) into three nodes (56 core, 250G memory, 10G network, SSD).
```
Comment (whitewum) [inline docs/manual-EN/3.build-develop-and-administration/3.deploy-and-administrations/server-administration/storage-service-administration/data-import/spark-writer.md]: 400k QPS

**Comment by tong-hao** [comment]
Which version of spark  that spark-writer support?
I have an error.

- run:
```
~/software/spark-1.6.2-bin-hadoop2.6/bin/spark-submit --class com.vesoft.nebula.tools.generator.v2.SparkClientGenerator --master local[2]   ./sst.generator-1.0.0-beta.jar  -c idmapping.conf -h -d
```

- Exception:
```

19/11/29 16:34:37 INFO SparkClientGenerator$: Nebula Addresses List(127.0.0.1:3699) for user:password
19/11/29 16:34:37 INFO SparkClientGenerator$: Connection Timeout 3000 Retry 3
19/11/29 16:34:37 INFO SparkClientGenerator$: Execution Retry 3 Interval Base 3000
19/11/29 16:34:37 INFO SparkClientGenerator$: Switch to idmapping
19/11/29 16:34:37 INFO SparkClientGenerator$: Spark Config Set spark.app.name:Spark Writer
19/11/29 16:34:37 INFO SparkClientGenerator$: Spark Config Set spark.cores.max:16
19/11/29 16:34:37 INFO SparkClientGenerator$: Spark Config Set spark.driver.cores:1
19/11/29 16:34:37 INFO SparkClientGenerator$: Spark Config Set spark.driver.maxResultSize:1G
Exception in thread "main" java.lang.NoSuchMethodError: org.apache.spark.network.util.JavaUtils.byteStringAs(Ljava/lang/String;Lorg/apache/spark/network/util/ByteUnit;)J
	at org.apache.spark.internal.config.ConfigHelpers$.byteFromString(ConfigBuilder.scala:67)
	at org.apache.spark.internal.config.ConfigBuilder$$anonfun$bytesConf$1.apply(ConfigBuilder.scala:235)
	at org.apache.spark.internal.config.ConfigBuilder$$anonfun$bytesConf$1.apply(ConfigBuilder.scala:235)
	at org.apache.spark.internal.config.TypedConfigBuilder.createWithDefault(ConfigBuilder.scala:143)
	at org.apache.spark.sql.internal.SQLConf$.<init>(SQLConf.scala:254)
	at org.apache.spark.sql.internal.SQLConf$.<clinit>(SQLConf.scala)
	at org.apache.spark.sql.internal.StaticSQLConf$.<init>(StaticSQLConf.scala:31)
	at org.apache.spark.sql.internal.StaticSQLConf$.<clinit>(StaticSQLConf.scala)
	at org.apache.spark.sql.SparkSession$Builder.enableHiveSupport(SparkSession.scala:867)
	at com.vesoft.nebula.tools.generator.v2.SparkClientGenerator$.main(SparkClientGenerator.scala:144)
	at com.vesoft.nebula.tools.generator.v2.SparkClientGenerator.main(SparkClientGenerator.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:731)
	at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:181)
	at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:206)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:121)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)

```

**Comment by Amber1990Zhang** [comment]
> Which version of spark that spark-writer support?
> I have an error.
> 
> * run:
> 
> ```
> ~/software/spark-1.6.2-bin-hadoop2.6/bin/spark-submit --class com.vesoft.nebula.tools.generator.v2.SparkClientGenerator --master local[2]   ./sst.generator-1.0.0-beta.jar  -c idmapping.conf -h -d
> ```
> 
> * Exception:
> 
> ```
> 
> 19/11/29 16:34:37 INFO SparkClientGenerator$: Nebula Addresses List(127.0.0.1:3699) for user:password
> 19/11/29 16:34:37 INFO SparkClientGenerator$: Connection Timeout 3000 Retry 3
> 19/11/29 16:34:37 INFO SparkClientGenerator$: Execution Retry 3 Interval Base 3000
> 19/11/29 16:34:37 INFO SparkClientGenerator$: Switch to idmapping
> 19/11/29 16:34:37 INFO SparkClientGenerator$: Spark Config Set spark.app.name:Spark Writer
> 19/11/29 16:34:37 INFO SparkClientGenerator$: Spark Config Set spark.cores.max:16
> 19/11/29 16:34:37 INFO SparkClientGenerator$: Spark Config Set spark.driver.cores:1
> 19/11/29 16:34:37 INFO SparkClientGenerator$: Spark Config Set spark.driver.maxResultSize:1G
> Exception in thread "main" java.lang.NoSuchMethodError: org.apache.spark.network.util.JavaUtils.byteStringAs(Ljava/lang/String;Lorg/apache/spark/network/util/ByteUnit;)J
> 	at org.apache.spark.internal.config.ConfigHelpers$.byteFromString(ConfigBuilder.scala:67)
> 	at org.apache.spark.internal.config.ConfigBuilder$$anonfun$bytesConf$1.apply(ConfigBuilder.scala:235)
> 	at org.apache.spark.internal.config.ConfigBuilder$$anonfun$bytesConf$1.apply(ConfigBuilder.scala:235)
> 	at org.apache.spark.internal.config.TypedConfigBuilder.createWithDefault(ConfigBuilder.scala:143)
> 	at org.apache.spark.sql.internal.SQLConf$.<init>(SQLConf.scala:254)
> 	at org.apache.spark.sql.internal.SQLConf$.<clinit>(SQLConf.scala)
> 	at org.apache.spark.sql.internal.StaticSQLConf$.<init>(StaticSQLConf.scala:31)
> 	at org.apache.spark.sql.internal.StaticSQLConf$.<clinit>(StaticSQLConf.scala)
> 	at org.apache.spark.sql.SparkSession$Builder.enableHiveSupport(SparkSession.scala:867)
> 	at com.vesoft.nebula.tools.generator.v2.SparkClientGenerator$.main(SparkClientGenerator.scala:144)
> 	at com.vesoft.nebula.tools.generator.v2.SparkClientGenerator.main(SparkClientGenerator.scala)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.lang.reflect.Method.invoke(Method.java:498)
> 	at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:731)
> 	at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:181)
> 	at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:206)
> 	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:121)
> 	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
> ```
@darionyaphet @critical27 
Please help,

**Comment by darionyaphet** [comment]
Spark 2.X is supported and we haved tested 2.2 and 2.4.

**Pull request merged**